### PR TITLE
added night-time scheduled builds

### PR DESCRIPTION
**Pull request by tyvision**
Signed-off-by: tyvision <antykushin@gmail.com>

nightly builds added

Signed-off-by: tyvision <antykushin@gmail.com>

### Description of the Change

Added implementation of nightly builds:
Scheduled at 23:00 UTC every day with the following parameters
- build for platforms:
    - Linux
    - MacOS
    - ARMv8
- without Doxygen
- without bindings

### Possible Drawbacks 

1) Builds for all branches will be triggered, but then stopped except for 'develop'.
2) as soon as nightly build is fully configured and managed by Jenkinsfile, if any of the nightly build parameter needs to be changed --> new PR should be opened

### CODE REVIEW

**Comment by l4l** [comment]
Maybe **nightly** -> **night-time**
I have an association of unstable packages for nightly builds (e.g firefox)

File: .jenkinsci/cancel-nightly-except-develop.groovy [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    def job = Jenkins.instance.getItemByFullName(jobname)
+
+    if (jobname =~ /^.*\/${job.name}$/) {
+        for (build in job.builds) {
+            if (!build.isBuilding()) { continue; }
+            if (buildnum == build.getNumber().toInteger()) { build.doStop(); }
+        }
+    }
+}
+return this
```
Comment (bakhtin) [inline .jenkinsci/cancel-nightly-except-develop.groovy]: new line

File: .jenkinsci/cancel-nightly-except-develop.groovy [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env groovy
```
Comment (bakhtin) [inline .jenkinsci/cancel-nightly-except-develop.groovy]: How much time does it take to go through existing branches and cancel all of them but develop?

File: Jenkinsfile [hunk 3]
```diff
@@ -41,19 +42,38 @@ pipeline {
...
+          if (BRANCH_NAME != "develop") {
+            if (params.Nightly) {
+                // Stop this job running if it is nightly but not the develop it should be
+                def tmp = load ".jenkinsci/cancel-nightly-except-develop.groovy"
+                tmp.cancelThisJob()
+            }
+            else {
+              // Stop same job running builds if it is commit/PR build and not triggered as nightly
+              def builds = load ".jenkinsci/cancel-builds-same-job.groovy"
+              builds.cancelSameJobBuilds()
```
Comment (bakhtin) [inline Jenkinsfile]: Please, do not cancel builds if they're already running. Let commit continue building and run nightly build alongside already running one.

File: .jenkinsci/cancel-nightly-except-develop.groovy (hunk 2 above)
Comment (tyvision) [reply .jenkinsci/cancel-nightly-except-develop.groovy]: could not estimate it, because it worked only for one branch by the moment

**Pull request merged**